lib/spack/spack/util/windows_registry.py

```diff
@@ -0,0 +1,262 @@
+        return self._keys
+
+    def gather_subkey_info(self):
+        """Composes all subjeys into a list for access"""
+        if self._keys:
+            return
+        sub_keys, _, _ = winreg.QueryInfoKey(self.handle)
+        for i in range(sub_keys):
+            sub_name = winreg.EnumKey(self.handle, i)
+            sub_handle = winreg.OpenKeyEx(self.handle, sub_name, access=winreg.KEY_READ)
+            self._keys.append(RegistryKey(os.path.join(self.name, sub_name), sub_handle))
+
+    def gather_value_info(self):
+        """Compose all values for this key into a dict of form value name: RegistryValue Object"""
+        if self._values:
+            return
+        _, values, _ = winreg.QueryInfoKey(self.handle)
+        for i in range(values):
+            value_name, value_data, _ = winreg.EnumValue(self.handle, i)
+            self._values[value_name] = RegistryValue(value_name, value_data, self.handle)
+
+    def get_subkey(self, sub_key):
+        """Returns subkey of name sub_key in a RegistryKey objects"""
+        return RegistryKey(
+            os.path.join(self.path, sub_key),
+            winreg.OpenKeyEx(self.handle, sub_key, access=winreg.KEY_READ),
+        )
+
+    def get_value(self, val_name):
+        """Returns value associated with this key in RegistryValue object"""
+        return RegistryValue(val_name, winreg.QueryValueEx(self.handle, val_name)[0], self.handle)
+
+
+class WindowsRegistry(object):
+    """
+    Interface to provide access, querying, and searching to Windows registry entries.
+    This class represents a single key entrypoint into the Windows registry
+    and provides an interface to this key's values, its subkeys, and those subkey's values.
+    This class cannot be used to move freely about the registry, only subkeys/values of
+    the root key used to instantiate this class.
+    """
+
+    def __init__(self, key, root_key=HKEY_CONSTANT("HKEY_CURRENT_USER")):
+        """Constructs a Windows Registry entrypoint to key provided
+        root_string should be an already open root key or an hkey constant if provided
+
+        Args:
+            key (str): registry key to provide root for registry key for this clas
+            root_string: Already open registry key or HKEY constant to provide access into
+                         the Windows registry. Registry access requires an already open key
+                         to get an entrypoint, the HKEY constants are always open, or an already
+                         open key can be used instead.
+        """
+        if not is_windows:
+            raise RuntimeError(
+                "Cannot instantiate Windows Registry class on non Windows platforms"
+            )
+        self.key = key
+        self.root = root_key
+        try:
+            self.reg = RegistryKey(
+                os.path.join(str(self.root), self.key),
+                winreg.OpenKeyEx(self.root.hkey, self.key, access=winreg.KEY_READ),
+            )
```

Comment:
Where possible, it is best to avoid relying on external services in initializers: could this be initialized with an already-constructed `RegistryKey`? That would make this easier to test (in isolation of the actual registry).

You can add a static function like:

```
class WindowsRegistry()
  @staticmethod
  from_string(key, root_key):
    return WindowsRegistry(RegistryKey(...))
```